feat(frontend): integrate member authentication with baodaozai Q&A feature

ericsen-tsai opened the pull request:
## Summary

This PR integrates member authentication with the baodaozai (報到仔) Q&A feature, enabling logged-in members to submit and persist their answers to both choice and essay questions. The implementation includes React Query for efficient data fetching and mutations, dynamic question selection based on member preferences, and proper answer persistence with create/update logic.

## Changes

- **Added React Query integration**: Installed `@tanstack/react-query` and `@tanstack/react-query-devtools` for state management and data fetching
- **Created React Query utilities**: Added `get-query-client.ts` with server/client query client setup and proper hydration configuration
- **Implemented answer API layer**: 
  - Created GraphQL queries and mutations for `post-choice-answer` and `post-essay-answer`
  - Added API functions for fetching, creating, and updating answers with proper authentication
- **Added React Query hooks**: Created custom hooks for querying and mutating choice and essay answers (`usePostChoiceAnswersQuery`, `usePostEssayAnswersQuery`, and their respective mutation hooks)
- **Implemented batch answer submission**: Created `use-batch-submit-answers.ts` hook that handles batch submission of answers with create/update logic based on existing answers
- **Updated article component**: 
  - Integrated with `useHydratedAuthStore` to access member information and authentication tokens
  - Replaced mock login check with real authentication state
  - Added dynamic question selection based on member's `essayQuestionCount` preference
  - Implemented `showBaodaozai` visibility logic based on post and member settings
  - Connected Q&A modal submission to batch answer submission hook
- **Added BaodaozaiVisibilitySetter component**: New component to control baodaozai visibility based on member/post settings
- **Updated GraphQL post query**: Added `showBaodaozai` field and pagination parameters (`postEssayQuestionsTake`, `postChoiceQuestionsTake`) for question fetching
- **Refactored layout structure**: Moved providers (QueryClientProvider, AuthProvider, HeaderProvider) into a new `Providers` component for better organization
- **Enhanced Q&A modal**: 
  - Updated `handlePass` to properly clear answers when skipping questions
  - Fixed answer state management for essay questions
  - Improved styling class ordering
- **Updated type definitions**: Added `id` field to `BaodaozaiEssayQuestion` and `BaodaozaiChoiceQuestion` types
- **Added constants**: Created `BAODAOZAI_QUESTION_COUNT` constant for question count management
- **Removed mock file**: Deleted `mock.ts` file that contained placeholder login state and questions

## Additional Notes

- The implementation uses React Query's `staleTime: Infinity` for answer queries to prevent unnecessary refetches, as answers are typically static once submitted
- Batch submission handles both create and update operations by checking existing answers before deciding which mutation to use
- The visibility logic follows a hierarchy: post-level `showBaodaozai` setting takes precedence, then falls back to member-level setting
- Question selection dynamically balances choice and essay questions based on the member's `essayQuestionCount` preference
- All API calls include proper error handling and logging using `@twreporter/errors`
- This PR is based on #811 

## Screenshot(s)

N/A

## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/QA-2948dbdca93280a0b86cece18b8738f4?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)

ericsen-tsai committed:
feat(frontend): add post choice and essay answer integration

ericsen-tsai committed:
feat(frontend): enhance post handling with batch submission and query integration

- Added batch submission functionality for post essay and choice answers.
- Introduced new hooks for managing post choice and essay answers using React Query.
- Updated GraphQL queries to support pagination for essay and choice questions.
- Refactored article component to utilize new hooks and streamline answer submission process.
- Created a Providers component for better context management across the application.

ericsen-tsai committed:
feat(frontend): add baodaozai visibility setter and sort qa modal classname

ericsen-tsai committed:
chore(frontend): add react query and devtools

ericsen-tsai commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -315,19 +300,27 @@ const Article = ({ post }: { post: NonNullable<GetPostQuery['post']> }) => {
+
+    const essayQuestions = (post.postEssayQuestions ?? []).slice(0, essayCount)
+    const choiceQuestions = (post.postChoiceQuestions ?? []).slice(
+      BAODAOZAI_QUESTION_COUNT - choiceCount,
+      BAODAOZAI_QUESTION_COUNT
```
This is the intentional behavior.

ericsen-tsai committed:
fix(api-gateway): use errors.helpers.wrap instead of errors.headers.wrap

ericsen-tsai commented on packages/api-gateway/src/auth-mini-app.ts:
```diff
@@ -12,14 +12,76 @@ const errors = _errors.default
+  try {
+    return JSON.parse(decodedString)
+  } catch (_err) {
+    throw errors.headers.wrap(
```
Fixed in [fix(api-gateway): use errors.helpers.wrap instead of errors.headers.wrap](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/0165d17f490eecda4ee89efcf7fc41a3c337f042)

nickhsine commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -297,19 +282,27 @@ const Article = ({ post }: { post: NonNullable<GetPostQuery['post']> }) => {
+    if (typeof post?.showBaodaozai === 'boolean' && !post.showBaodaozai) {
+      return false
+    }
+    if (typeof member?.showBaodaozai === 'boolean') return member.showBaodaozai
```
nitpick:
可以改寫成
```
if (post?.showBaodaozai === true && member?.showBaodaozai === true) {
  return true
}

return false
```
嗎？

ericsen-tsai replied:
Fixed in [fix(frontend): select answer by post slug](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/513495e39fccdfc5f565e24df0c7a32a9c96e925)

nickhsine commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -297,19 +282,27 @@ const Article = ({ post }: { post: NonNullable<GetPostQuery['post']> }) => {
+  })()
 
-    const candidateQuestions = [...choiceQuestions]
+  const essayQuestionCount = member?.essayQuestionCount ?? 3
```
question:
昨天會議討論，預設的思辨題是不是一題？

ericsen-tsai replied:
Yes, fixed in [fix(frontend): select answer by post slug](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/513495e39fccdfc5f565e24df0c7a32a9c96e925)

nickhsine commented on packages/frontend/src/api/post-essay-answer.ts:
```diff
@@ -0,0 +1,67 @@
+      authToken: accessToken,
+    }
+  )
+  return response?.data?.data?.postEssayAnswers ?? []
```
question:
`getPostEssayAnswersByMemberId` 預設是拿回某一個 member 的所有 essay answers 嗎？

ericsen-tsai replied:
It should be get the "answers for a specific post for a specific member". Fixed in [fix(frontend): select answer by post slug](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/6df61855cf2c531d7311f3df1ceaad6d1c8ce65f)

nickhsine commented on packages/frontend/src/api/post-choice-answer.ts:
```diff
@@ -0,0 +1,67 @@
+      authToken: accessToken,
+    }
+  )
+  return response?.data?.data?.postChoiceAnswers ?? []
```
question:
`getPostChoiceAnswersByMemberId` 預設是拿回某一個 user 的所有 choice answers，而不是「拿某一個 user，針對某一文章的所有回答嗎？」

ericsen-tsai replied:
Yes, fixed in [fix(frontend): select answer by post slug](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/6df61855cf2c531d7311f3df1ceaad6d1c8ce65f).

ericsen-tsai committed:
fix(frontend): add default essay count and simplify show baodaozai logic

ericsen-tsai committed:
fix(frontend): select answer by post slug

nickhsine commented on packages/frontend/src/api-utils/react-query/get-query-client.ts:
```diff
@@ -0,0 +1,31 @@
+  } else {
+    if (!browserQueryClient) browserQueryClient = makeQueryClient()
+    return browserQueryClient
+  }
```
好奇想問問，server side 和 client side 在 initialize client 的方式有別，是有什麼考量嗎？

ericsen-tsai replied:
We want to create new query client on server to prevent polluting other request.
However, in client, we just need to keep the `queryClient` as one instance to keep the cache.

ref: https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr#initial-setup

nickhsine commented on packages/frontend/src/api/graphql/post-choice-answer.ts:
```diff
@@ -0,0 +1,48 @@
+      }
+      member {
+        id
+      }
```
note:
因為 https://github.com/kids-reporter/kids-reporter-monorepo/pull/823 調整了 createPostChoiceAnswer 和 createPostEssayAnswer mutations，如果此 PR 先 merge 的話， PR 823 需要調整此 mutation。

ericsen-tsai replied:
Modified in [refactor(frontend): remove member field from GraphQL mutations for post answers](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/7170dc172df0de1c2bfec8cc485cb0e87d374c91)

nickhsine commented on packages/frontend/src/api/graphql/post-choice-answer.ts:
```diff
@@ -0,0 +1,48 @@
+      id
+      member {
+        id
+      }
```
同上的 note。

ericsen-tsai replied:
Modified in [refactor(frontend): remove member field from GraphQL mutations for post answers](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/7170dc172df0de1c2bfec8cc485cb0e87d374c91)

nickhsine commented on packages/frontend/src/api/graphql/post-essay-answer.ts:
```diff
@@ -0,0 +1,42 @@
+      }
+      member {
+        id
+      }
```
同上個 note。

ericsen-tsai replied:
Modified in [refactor(frontend): remove member field from GraphQL mutations for post answers](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/7170dc172df0de1c2bfec8cc485cb0e87d374c91)

nickhsine commented on packages/frontend/src/api/graphql/post-essay-answer.ts:
```diff
@@ -0,0 +1,42 @@
+      id
+      member {
+        id
+      }
```
同上個 note。

ericsen-tsai replied:
Modified in [refactor(frontend): remove member field from GraphQL mutations for post answers](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/7170dc172df0de1c2bfec8cc485cb0e87d374c91)

nickhsine commented on packages/frontend/src/app/(sticky-header)/_components/article/hooks/use-batch-submit-answers.ts:
```diff
@@ -0,0 +1,179 @@
+                const answerId =
+                  essayAnswers?.find(
+                    (answer) => answer.question?.id === question.id
+                  )?.id ?? ''
```
nitpick:
上述判斷好像可以不用 for loop 兩次？

```
const idx = existingEssayQuestionIds.indexOf(question.id)
if (idx > -1) {
  const answerId = essayAnswers[idx]
  // update post essay answer
} else {
  // create post essay answer
}
```

ericsen-tsai replied:
Fixed in [fix(frontend): improve answer submission logic in batch processing](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/58ee0bc0dd2cb2ba9107d945063b166101ee1eec)

nickhsine commented on packages/frontend/src/app/(sticky-header)/_components/article/hooks/use-batch-submit-answers.ts:
```diff
@@ -0,0 +1,179 @@
+            const msg = errors.helpers.printAll(err, {
+              withStack: true,
+              withPayload: true,
+            })
```
suggestion:
```
            const msg = errors.helpers.printAll(err, {
              withStack: true,
              withPayload: true,
            }, 0, 0)
```

最後一個參數設定成 0，可以印出完整的 stack trace，不然預設只會印出 stack trace 的第一行。

reference: https://github.com/twreporter/twreporter-npm-packages/blob/master/packages/errors/src/helpers/print.js#L46

ericsen-tsai replied:
Fixed in [fix(frontend): improve answer submission logic in batch processing](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/58ee0bc0dd2cb2ba9107d945063b166101ee1eec)

nickhsine commented on packages/frontend/src/services/call-baodaozai/components/qa-modal/index.tsx:
```diff
@@ -43,9 +43,13 @@ function QAModal({ questions, onClose, onSubmit, isOpen }: QAModalProps) {
+  const handlePass = useCallback(
+    (questionIndex: number) => {
+      setCurrentModalStepIndex(currentModalStepIndex + 2)
+      setAnswers((prev) => ({ ...prev, [questionIndex]: '' }))
```
question:
如果是跳過題目，目前的機制是會產生空白答案，還是會不產生答案呢？

ericsen-tsai replied:
It won't create new record. [use-batch-submit-answers](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/files#diff-2bcd79c6be84a20973d47078ae2419010102cdfe77d148869a24afb1221f1f26R83)
But I clarify the logic in [fix(frontend): improve answer submission logic in batch processing](https://github.com/kids-reporter/kids-reporter-monorepo/pull/817/commits/58ee0bc0dd2cb2ba9107d945063b166101ee1eec) to just delete the property

nickhsine reviewed (commented):
LGTM 👍 👍  
Nice nice.

ericsen-tsai committed:
chore(eslint): update no-unused-vars rule to ignore leading underscores for variables and arguments

ericsen-tsai committed:
fix(frontend): improve answer submission logic in batch processing

- Refactored answer retrieval for essay and choice questions to use index-based access instead of searching by ID.
- Enhanced error logging to include additional parameters for better debugging.
- Updated answer state management in QAModal to remove answers instead of setting them to empty strings.

ericsen-tsai committed:
refactor(frontend): remove member field from GraphQL mutations for post answers

- Eliminated the member field from CREATE and UPDATE mutations for both post choice and essay answers.
- Updated related batch submission logic to reflect the removal of member connections.

ericsen-tsai committed:
fix(frontend): update baodaozai visibility logic and improve answer handling

- Adjusted the logic for displaying baodaozai based on user login status.
- Set default essay question count to zero for non-logged-in users.
- Refactored answer handling in QAModal to use index-based access for correct answers to submit correct answer index.

ericsen-tsai commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -292,14 +292,22 @@ const Article = ({
 
   const showBaodaozai = (() => {
-    if (post?.showBaodaozai === true && member?.showBaodaozai === true) {
+    if (post?.showBaodaozai === true && !isLogin) {
+      return true
+    }
+    if (
+      post?.showBaodaozai === true &&
+      isLogin &&
+      member?.showBaodaozai === true
+    ) {
```
@nickhsine 
The former logic is incorrect because the non-login user should see the baodaozai if the config is open. I've fixed this one.

ericsen-tsai commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -292,14 +292,22 @@ const Article = ({
 
-  const essayQuestionCount =
-    member?.essayQuestionCount ?? BAODAOZAI_DEFAULT_ESSAY_QUESTION_COUNT
+  const essayQuestionCount = isLogin
+    ? (member?.essayQuestionCount ?? BAODAOZAI_DEFAULT_ESSAY_QUESTION_COUNT)
+    : 0
```
@nickhsine 
This is also the fix to handle non-loggedin user.
